system/src/system_network_internal.h

```diff
@@ -622,6 +624,24 @@ class ManagedNetworkInterface : public NetworkInterface
         }
     }
 
+    bool isOn() override
+    {
+#if PLATFORM_ID == PLATFORM_ELECTRON // Electron
+        return !WLAN_INITIALIZED || SPARK_WLAN_STARTED;
+#else
+        return SPARK_WLAN_STARTED;
+#endif
+    }
+
+    bool isOff() override
+    {
+#if PLATFORM_ID == PLATFORM_ELECTRON // Electron
+        return WLAN_INITIALIZED && !SPARK_WLAN_STARTED;
```

Comment:
Probably let's remove conditional for Electron (unless we know for a fact that BCM on Photon is off on boot).

This looks okay, because technically on Electron `!WLAN_INITIALIZED && !SPARK_WLAN_STARTED` means that the state is unknown. Ideally we'd query the modem for its state, but returning false for both `isOn()` and `isOff()` might also be valid.

Reply:
If `isOff()` intends to indicate the power state, then we should return return `return WLAN_INITIALIZED && !SPARK_WLAN_STARTED;` for all Gen2 platforms here.

Reply:
I was still quite confused why isOff() would `return WLAN_INITIALIZED && ...` since it should indicate we are powered off (!SPARK_WLAN_STARTED) and de-initialized (!WLAN_INITIALIZED).  But when I looked through system_network_internal.h I don't see off() setting WLAN_INITIALIZED = 0;  So isOff() would still be false after calling off() with returning `!WLAN_INITIALIZED &&` .  This seems like an error in system_network_internal.h and we should be setting WLAN_INITIALIZED = 0 in off(), which would then allow us to `return !WLAN_INITIALIZED && !SPARK_WLAN_STARTED;` for isOff().

Reply:
I know why you are confused, as you were expecting the `WLAN_INITIALIZED` flag to indicating the low-level modem state. While the truth is that the `WLAN_INITIALIZED` was introduced in previous PR https://github.com/particle-iot/device-os/pull/2197, which is to fix the power leakage issue on Electron on bootup. The context is that when the Electron is powered up the modem does consuming high current if user application didn't touch any network API and then go to sleep. So this flag only indicates whether the modem is initialized at least once so that the modem power state resides in a well known state and you can see that it is set only in the `on()` function in system_network_internal.h. 

That's why `isOff()` should return `WLAN_INITIALIZED && !SPARK_WLAN_STARTED`, as if `WLAN_INITIALIZED` is not set, we consider that the modem is powered up. As long as the `WLAN_INITIALIZED` is set, the modem state is tracked by `SPARK_WLAN_STARTED`, which is changed along with the `on()`/`off()` API.

Reply:
If `WLAN_INITIALIZED` is not set, the modem is in an unknown state, neither `isOn()` nor `isOff()` should return true.

Reply:
Ok thanks for the explanation.  `WLAN_INITIALIZED` should be commented in firmware with this understanding of how it should be used and what it means, because just looking at it you can't tell that it should not be reset if the modem is uninitialized.  We possibly shouldn't be setting `WLAN_INITIALIZED` back to 0 in `on()` if it's meant to be a static flag that only gets set when the modem is first turned on.  You can't say calling on() twice would turn off the modem, and the second time through it might fail the on_now() call and report that it's off `WLAN_INITIALIZED = 0`, when it might have been turned on the first time and is still on.

Reply:
Yeah. I agree with you. @avtolstoy Any reason why we were adding this in: https://github.com/particle-iot/device-os/commit/c6fc6c3da8bef1fe47cb5afadc2056d980109b48#diff-f61a83197bcd7fe3529f7ca6e612a99d1222bf9d8080af2b0c4c7693fb9bf95dR620?

Reply:
The PR that contains that commit explains why we are doing this. It is reset to keep track of whether the power-on sequence actually succeeded and if not retry it on the subsequent attempts. When we request the modem to turn on we set `SPARK_WLAN_STARTED` to 1. There is a lot of other code that relies on this condition, so we cannot use it to keep track of whether we need to retry or not. So, `WLAN_INITIALIZED` is used for that and will be reset to 0 while powering on or after a failed power on attempt to indicate that we are in an unknown state. This plus `SPARK_WLAN_STARTED == 1` allows us to retry power-on.

Reply:
That makes sense. Thanks for the explanation! 👍